Error when compiling with  Windows 10 SDK 10.0.14393.0

Almost two thousand errors complaining about `constexpr`.

>
c:\program files (x86)\windows kits\10\include\10.0.14393.0\um\shlwapi.h(1764): error C3249: illegal statement or sub-expression for 'constexpr' function [C:\DEV\GIT\Forks\coreclr\bin\obj\Windows_NT.x64.Checked\src\debug\ee\dac\cordbee_dac.vcxproj]

Still have an issue, is different though
```
coreclr\src\vm\comcallablewrapper.cpp(2388): 
 error C4839: non-standard use of class 'NewCCWHolder' as an argument to a variadic function 
 [coreclr\bin\obj\Windows_NT.x64.Checked\src\vm\wks\cee_wks.vcxproj]
coreclr\src\vm\comcallablewrapper.cpp(2388): 
 error C2280: 'NewCCWHolder::NewCCWHolder(const NewCCWHolder &)': 
 attempting to reference a deleted function 
 [coreclr\bin\obj\Windows_NT.x64.Checked\src\vm\wks\cee_wks.vcxproj]
```

@pgavlin, at least the hello world with SDK headers compile:

```cpp
#include <stdio.h>
#include <stdlib.h>
#include <windows.h>
#include "c:\program files (x86)\windows kits\10\include\10.0.14393.0\um\shlwapi.h"

int main(int argc, char **argv)
{
        fprintf(stderr,"\nHello World\n");
        return 0;
}
```

but compiling CoreCLR fails with:

> c:\program files (x86)\windows kits\10\include\10.0.14393.0\um\shlwapi.h(1764): error C3249: illegal statement or sub-expression for 'constexpr' function (compiling source file C:\Users\Peter\Source\Repos\coreclr\src\classlibnative\bcltype\arrayhelpers.cpp) [C:\Users\Peter\Source\Repos\coreclr\bin\obj\Windows_NT.x64.Debug\src\classlibnative\bcltype\bcltype.vcxproj]

This is VS2015 Update 3 with SDK 14393.

Then I grabbed latest VC bits and followed @benaadams' approach:

```powershell
nuget install VisualCppTools -source http://vcppdogfooding.azurewebsites.net/nuget/ -Prerelease

# backup
mkdir C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\temp

mv C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\bin C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\temp

mv C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\include C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\temp

mv C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\lib C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC\lib

# copy latest VCR
cp -r VisualCppTools.14.0.24405-Pre\lib\native\* C:\Program Files (x86)\Microsoft Visual Studio 14.0\VC
```

and then in coreclr dir `.\build.cmd` and it didn't complain about constexpr.

@benaadams, `error C4839` you encountered with VC toolset grabbed from NuGet: https://github.com/dotnet/coreclr/issues/6642#issuecomment-238268603 is fixed by #6665 :tada:

@jkotas, it gives the same error:

>
c:\program files (x86)\windows kits\10\include\10.0.14393.0\um\shlwapi.h(1764): error C3249: illegal statement or sub-expression for 'constexpr' function

```diff
C:\Users\Peter\Source\Repos\coreclr [master ≡ +0 ~1 -0 !]> git diff
diff --git a/src/inc/debugreturn.h b/src/inc/debugreturn.h
index cc5bdeb..2a012fb 100644
--- a/src/inc/debugreturn.h
+++ b/src/inc/debugreturn.h
@@ -63,8 +63,13 @@
 //
 class __SafeToReturn {
 public:
+#if _MSC_VER >= 1900
+    static constexpr int safe_to_return() {return 0;};
+    static constexpr int used() {return 0;};
+#else
     static int safe_to_return() {return 0;};
     static int used() {return 0;};
+#endif
 };

 class __YouCannotUseAReturnStatementHere {
```

--- a/src/inc/debugreturn.h
+++ b/src/inc/debugreturn.h
@@ -27,7 +27,9 @@
 
 #else // !_PREFAST_
 
-#ifdef _DEBUG
+// This is disabled in VS2015 Update 3 and earlier because only C++11 constexpr is supported,
+// which doesn't allow the use of 'if' statements within the body of a constexpr function.
+#if defined(_DEBUG) && (!defined(_MSC_FULL_VER) || _MSC_FULL_VER > 190024210)
 
 // Code to generate a compile-time error if return statements appear where they
 // shouldn't.
@@ -108,15 +110,17 @@ typedef __SafeToReturn __ReturnOK;
 #define DEBUG_OK_TO_RETURN_BEGIN(arg) { typedef __SafeToReturn __ReturnOK; if (0 && __ReturnOK::used()) { } else {
 #define DEBUG_OK_TO_RETURN_END(arg) } }
 
-#else // !_DEBUG
+#else // defined(_DEBUG) && (!defined(_MSC_FULL_VER) || _MSC_FULL_VER > 190024210)
 
-#define DEBUG_ASSURE_NO_RETURN_BEGIN(arg)
-#define DEBUG_ASSURE_NO_RETURN_END(arg)
+#define DEBUG_ASSURE_SAFE_TO_RETURN TRUE
 
-#define DEBUG_OK_TO_RETURN_BEGIN(arg)
-#define DEBUG_OK_TO_RETURN_END(arg)
+#define DEBUG_ASSURE_NO_RETURN_BEGIN(arg) {
+#define DEBUG_ASSURE_NO_RETURN_END(arg) }
 
-#endif // !_DEBUG
+#define DEBUG_OK_TO_RETURN_BEGIN(arg) {
+#define DEBUG_OK_TO_RETURN_END(arg) }
+
+#endif // defined(_DEBUG) && (!defined(_MSC_FULL_VER) || _MSC_FULL_VER > 190024210)
 
 #endif // !_PREFAST_
 

